core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -459,15 +462,21 @@ public void close()
     private static class SpoolingExchangeDataSource
             implements ExchangeDataSource
     {
-        private final ExchangeSource exchangeSource;
+        // This field is not final to allow releasing the memory retained by the ExchangeSource instance.
+        // It is modified (assigned to null) when the ExchangeOperator is closed.
+        // It doesn't have to be declared as volatile as the nullification of this variable doesn't have to be immediately visible to other threads.
+        // However since close can be called at any moment this variable has to be accessed in a safe way (avoiding "check-then-use").
+        private ExchangeSource exchangeSource;
```

Comment:
There two separate issues here. One improving GC and another is ensuring memory pool accounting is fine. It would be nice to have separated commits for them.